contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,340 @@
+    AsUser string
+    // Impersonate-Group header value
+    AsGroups string
+    // Object reference this request is targeted at
+    Object ObjectReference
+    // Response status code are returned by the server
+    ResponseStatusCode int
+    // Error response, if ResponseStatusCode >= 400
+    ResponseErrorMessage string
+
+    // below fields are filled at RequestObject level and higher:
+
+    // RequestObject logged before admission (json format)
+    RequestObject runtime.Unstructured
+    // Response object in json format
+    ResponseObject runtime.Unstructured
+
+    // below fields are filled at StorageObject level and higher:
+
+    // Object value before modification (will be empty when creating new object)
+    OldObject runtime.Object
+    // Object value after modification (will be empty when removing object)
+    NewObject runtime.Object
+    // Difference between old and new object
+    Patch     []byte
+}
+```
+
+### Policy
+
+The audit policy determines what audit event is generated for a given request. The policy is configured by the cluster administrator. Here is a sketch of the policy API:
+
+```go
+type Policy struct {
+    // Rules specify the audit Level a request should be recorded at.
+    // A request may match multiple rules, in which case the FIRST matching rule is used.
+    // The default audit level is None, but can be overridden by a catch-all rule at the end of the list.
+    Rules []PolicyRule
+
+    // Discussed under Filters section.
+    Filters []Filter
+}
+
+// Based off the RBAC PolicyRule
+type PolicyRule struct {
+    // Required. The Level that requests matching this rule are recorded at.
+    Level Level
+
+    // The users (by authenticated user name) this rule applies to.
+    // An empty list implies every user.
+    Users []string
+    // The user groups this rule applies to. If a user is considered matching
+    // if the are a member of any of these groups
+    // An empty list implies every user group.
+    UserGroups []string
+
+    // The verbs that match this rule.
+    // An empty list implies every verb.
+    Verbs []string
+
+    // Rules can apply to API resources (such as "pods" or "secrets"),
+    // non-resource URL paths (such as "/api"), or neither, but not both.
+    // If neither is specified, the rule is treated as a default for all URLs.
+
+    // APIGroups is the name of the APIGroup that contains the resources ("" for core).
+    // If multiple API groups are specified, any action requested against one of the
+    // enumerated resources in any API group will be allowed.
+    // Any empty list implies every group.
+    APIGroups []string
+    // Namespaces that this rule matches.
+    // This field should be left empty if specifying non-namespaced resources.
+    // Any empty list implies every namespace.
+    Namespaces []string
+    // Resources is a list of resource types this rule applies to.
+    // Any empty list implies every resource type.
+    Resources []string
```

Comment:
I like David's approach. I think it's the most flexible, while cutting out some of the redundancy. I'll make that change.